File: workers/music-gateway/src/providers/qobuz-api.ts
import { md5Hex } from "../lib/md5";
import { normalizePublicHttpsUrl } from "../lib/safe-stream-url";
import { inferBitrateKbps, isHiResSignal } from "../lib/stream-quality";
import { qobuzArtistLine } from "../lib/qobuz-artists";
import { splitArtistCredits } from "../lib/artist-credits";
import type { Env, GatewayTrack, StreamResult } from "../types";
import { fetchJson, fetchText } from "./shared";

const UA = "VANTA-MusicGateway/2.0";
const QOBUZ_API_BASE = "https://www.qobuz.com/api.json/0.2";

interface QobuzCreds {
  appId: string;
  appSecret: string;
}

let cachedCreds: QobuzCreds | null = null;

export async function getQobuzCredentials(): Promise<QobuzCreds | null> {
  if (cachedCreds) return cachedCreds;

  const scraped = await scrapeQobuzOpenCredentials();
  if (scraped) {
    cachedCreds = scraped;
    return scraped;
  }

  return null;
}

async function scrapeQobuzOpenCredentials(): Promise<QobuzCreds | null> {
  const shell = await fetchText("https://open.qobuz.com/track/1", {
    headers: { "User-Agent": UA, Accept: "text/html" },
  });
  if (!shell) return null;

  const scriptMatch = shell.match(/<script[^>]+src="([^"]+\/js\/main\.js|\/resources\/[^"]+\/js\/main\.js)"/i);
  if (!scriptMatch?.[1]) return null;

  let bundleUrl = scriptMatch[1];
  if (bundleUrl.startsWith("/")) bundleUrl = `https://open.qobuz.com${bundleUrl}`;

  const bundle = await fetchText(bundleUrl, { headers: { "User-Agent": UA } });
  if (!bundle) return null;

  const configMatch = bundle.match(/app_id:"(\d{9})",app_secret:"([a-f0-9]{32})"/);
  if (!configMatch) return null;

  return { appId: configMatch[1], appSecret: configMatch[2] };
}

export function isQobuzSampleUrl(url: string): boolean {
  try {
    const parsed = new URL(url);
    const range = parsed.searchParams.get("range");
    if (!range) return false;
    const match = range.match(/^(\d+)-(\d+)$/);
    if (!match) return false;
    const start = Number(match[1]);
    const end = Number(match[2]);
    return end - start <= 40;
  } catch {
    return false;
  }
}

function qobuzSignature(path: string, params: Record<string, string>, timestamp: string, secret: string): string {
  const normalizedPath = path.replace(/^\/+|\/+$/g, "").replace(/\//g, "");
  const keys = Object.keys(params)
    .filter((k) => !["app_id", "request_ts", "request_sig"].includes(k))
    .sort();
  let payload = normalizedPath;
  for (const key of keys) {
    payload += key + params[key];
  }
  payload += timestamp + secret;
  return md5Hex(payload);
}

async function qobuzSignedGet(
  path: string,
  params: Record<string, string>,
  auth?: { userAuthToken?: string }
): Promise<unknown | null> {
  const creds = await getQobuzCredentials();
  if (!creds) return null;
  const timestamp = String(Math.floor(Date.now() / 1000));
  const signedParams = { ...params };
  if (auth?.userAuthToken) signedParams.user_auth_token = auth.userAuthToken;
  const query = new URLSearchParams({ ...signedParams, app_id: creds.appId, request_ts: timestamp });
  const sig = qobuzSignature(path, Object.fromEntries(query.entries()), timestamp, creds.appSecret);
  query.set("request_sig", sig);

  const url = `${QOBUZ_API_BASE}/${path.replace(/^\/+/, "")}?${query.toString()}`;
  const headers: Record<string, string> = {
    "User-Agent": UA,
    Accept: "application/json",
    "X-App-Id": creds.appId,
  };
  if (auth?.userAuthToken) headers["X-User-Auth-Token"] = auth.userAuthToken;
  return fetchJson(url, { headers });
}

export async function searchQobuzPublic(query: string, limit = 25): Promise<GatewayTrack[]> {
  const payload = (await qobuzSignedGet("track/search", { query, limit: String(limit) })) as {
    tracks?: { items?: Array<Record<string, unknown>> };
  } | null;

  return (payload?.tracks?.items ?? []).map((item) => mapQobuzTrack(item));
}

function mapQobuzTrack(item: Record<string, unknown>): GatewayTrack {
  const id = String(item.id ?? "");
  const bitDepth = typeof item.maximum_bit_depth === "number" ? item.maximum_bit_depth : 16;
  const sampleRate = typeof item.maximum_sampling_rate === "number" ? item.maximum_sampling_rate : 44.1;
  const album = item.album as { title?: string; id?: string | number; image?: { large?: string }; release_date_original?: string; release_date_stream?: string } | undefined;
  const audioQuality = `${bitDepth}-bit / ${sampleRate} kHz FLAC`;
  const channelCount = typeof item.maximum_channel_count === "number" ? item.maximum_channel_count : 2;
  const credits = splitArtistCredits(qobuzArtistLine(item));

  return {
    id,
    title: String(item.title ?? "Unknown"),
    artist: credits.primary || qobuzArtistLine(item),
    featuredArtists: credits.featured.length > 0 ? credits.featured : undefined,
    album: album?.title ? String(album.title) : undefined,
    albumId: album?.id != null ? String(album.id) : undefined,
    artworkURL: album?.image?.large,
    duration: typeof item.duration === "number" ? item.duration : undefined,
    releaseYear: parseReleaseYear(album?.release_date_original ?? album?.release_date_stream),
    trackNumber: typeof item.track_number === "number" ? item.track_number : undefined,
    discNumber: typeof item.media_number === "number" ? item.media_number : undefined,
    audioQuality: `${bitDepth}-bit / ${sampleRate} kHz FLAC`,
    isrc: typeof item.isrc === "string" ? item.isrc : undefined,
    format: "flac",
    explicit: Boolean(item.parental_warning),
    provider: "qobuz",
    qobuz_id: id,
    isDolbyAtmos: false,
    isSpatialAudio: false,
    isSurround: channelCount > 2,
    isHiRes: typeof item.hires_streamable === "boolean" ? item.hires_streamable : isHiResSignal(audioQuality, "flac", sampleRate, bitDepth),
  };
}

function parseReleaseYear(value: string | undefined): number | undefined {
  const match = value?.match(/^(\d{4})/);
  if (!match) return undefined;
  const year = Number(match[1]);
  return year >= 1900 && year <= new Date().getUTCFullYear() + 1 ? year : undefined;
}

export async function lookupQobuzTrackByIsrc(isrc: string): Promise<string | null> {
  const payload = (await qobuzSignedGet("track/search", { query: isrc, limit: "1" })) as {
    tracks?: { items?: Array<{ id?: number | string }> };
  } | null;
  const id = payload?.tracks?.items?.[0]?.id;
  return id != null ? String(id) : null;
}

export interface QobuzTrackMeta {
  id: string;
  title: string;
  artist: string;
  album?: string;
  isrc?: string;
  durationSec?: number;
}

/** Public Qobuz track metadata (no auth) — used to map a Qobuz id to other catalogs. */
export async function getQobuzTrackMeta(trackId: string): Promise<QobuzTrackMeta | null> {
  const id = String(trackId).trim();
  if (!id) return null;
  const payload = (await qobuzSignedGet("track/get", { track_id: id })) as {
    id?: number | string;
    title?: string;
    isrc?: string;
    duration?: number;
    performer?: { name?: string };
    performers?: string;
    album?: { title?: string };
  } | null;
  if (!payload?.title) return null;
  return {
    id: payload.id != null ? String(payload.id) : id,
    title: payload.title,
    artist: qobuzArtistLine(payload as Record<string, unknown>),
    album: payload.album?.title,
    isrc: payload.isrc,
    durationSec: typeof payload.duration === "number" ? payload.duration : undefined,
  };
}

export async function streamQobuzByoa(
  env: Env,
  trackId: string,
  quality: string,
  request: Request
): Promise<StreamResult | null> {
  const token = withByoaQobuzToken(env, request);
  if (!token) return null;
  return streamQobuzAuthenticated(env, trackId, quality, request);
}

/**
 * Community token path: uses a shared community Qobuz account token
 * (SpotiFLAC "qbz-a" style) so `format_id 27`/`7` hi-res FLAC can be
 * requested even when the operator has not configured their own QOBUZ_AUTH_TOKEN.
 * Falls back to anonymous quality sets when no community token is present.
 */
export async function streamQobuzPublic(
  env: Env,
  trackId: string,
  quality: string
): Promise<StreamResult | null> {
  if (!trackId.trim()) return null;

  const communityToken = env.QOBUZ_COMMUNITY_TOKEN?.trim();

  // With a community token we can request hi-res (format_id 27) and 7 the
  // same way the SpotiFLAC "community qbz-a" source does.
  const formatIds =
    quality === "16" ? ["6", "5"] : communityToken ? ["27", "7", "6", "5"] : ["7", "6", "5"];

  for (const formatId of formatIds) {
    const payload = await qobuzSignedGet(
      "track/getFileUrl",
      {
        track_id: trackId.trim(),
        format_id: formatId,
        intent: "stream",
      },
      { userAuthToken: communityToken ?? undefined }
    );
    const stream = qobuzFileUrlToStream(payload, quality);
    if (stream) return stream;
  }
  return null;
}

function byoaQobuzToken(request: Request): string | null {
  return request.headers.get("X-Qobuz-Token")?.trim() || request.headers.get("x-qobuz-token")?.trim() || null;
}

export function withByoaQobuzToken(env: Env, request: Request): string | null {
  return byoaQobuzToken(request) || env.QOBUZ_AUTH_TOKEN?.trim() || null;
}

/** Account stream via track/getFileUrl when the operator has configured a Qobuz session or BYOA header. */
export async function streamQobuzAuthenticated(
  env: Env,
  trackId: string,
  quality: string,
  request?: Request
): Promise<StreamResult | null> {
  const token = (request ? withByoaQobuzToken(env, request) : env.QOBUZ_AUTH_TOKEN?.trim()) || null;
  if (!token || !trackId.trim()) return null;

  const formatIds = quality === "16" ? ["6", "5"] : ["27", "7", "6", "5"];
  for (const formatId of formatIds) {
    const payload = await qobuzSignedGet(
      "track/getFileUrl",
      {
        track_id: trackId.trim(),
        format_id: formatId,
        intent: "stream",
      },
      { userAuthToken: token }
    );
    const stream = qobuzFileUrlToStream(payload, quality);
    if (stream) return stream;
  }
  return null;
}

export function qobuzFileUrlToStream(payload: unknown, quality: string): StreamResult | null {
  if (!payload || typeof payload !== "object") return null;
  const record = payload as Record<string, unknown>;
  const url = typeof record.url === "string" ? normalizePublicHttpsUrl(record.url) : null;
  if (!url) return null;
  if (isQobuzSampleUrl(url)) {
    console.warn(
      "VANTA_QOBUZ_SAMPLE_REJECTED",
      JSON.stringify({ urlHost: (() => { try { return new URL(url).host; } catch { return null; } })(), reason: "range-limited preview (no user_auth_token)" })
    );
    return null;
  }

  const mimeType = typeof record.mime_type === "string" ? record.mime_type : "audio/flac";
  const format = mimeType.includes("flac") ? "flac" : mimeType.replace(/^audio\//, "");
  const bitDepth = typeof record.bit_depth === "number" ? record.bit_depth : undefined;
  const samplingRate =
    typeof record.sampling_rate === "number"
      ? record.sampling_rate >= 1000
        ? record.sampling_rate / 1000
        : record.sampling_rate
      : undefined;
  const qualityLabel =
    bitDepth && samplingRate
      ? `${bitDepth}-bit / ${samplingRate} kHz ${format.toUpperCase()}`
      : format === "flac" ? "Lossless FLAC" : format.toUpperCase();

  return {
    url,
    streamUrl: url,
    format,
    quality: qualityLabel,
    mimeType,
    bitrateKbps: inferBitrateKbps(qualityLabel, format),
    bitDepth,
    sampleRateHz: samplingRate != null ? (samplingRate >= 1000 ? samplingRate : Math.round(samplingRate * 1000)) : undefined,
    provider: "qobuz",
    isDolbyAtmos: false,
    isSpatialAudio: false,
    isSurround: false,
    isHiRes: isHiResSignal(qualityLabel, format, samplingRate, bitDepth),
  };
}

File: vps-backend/qobuz_api.py
"""Qobuz API client for authenticated, full-track stream resolution."""
import hashlib
import os
import re
import time
from typing import Any, Optional
from urllib.parse import parse_qs, urlparse

import httpx

SEARCH_URL = "https://www.qobuz.com/api.json/0.2/track/search"
STREAM_URL = "https://www.qobuz.com/api.json/0.2/track/getFileUrl"
CREDENTIAL_URL = "https://open.qobuz.com/track/1"

QUALITY_FORMATS = {
    "hi_res": [27, 7, 6],
    "hi-res": [27, 7, 6],
    "hi_res_lossless": [27, 7, 6],
    "24": [27, 7, 6],
    "lossless": [6],
    "16": [6],
    "flac": [6],
    "16bit": [6],
    "mp3": [5],
    "low": [5],
}


class QobuzClient:
    def __init__(self, client: httpx.AsyncClient):
        self.client = client
        self.app_id = os.environ.get("QOBUZ_APP_ID", "").strip()
        self.app_secret = os.environ.get("QOBUZ_APP_SECRET", "").strip()
        self._ready = bool(self.app_id and self.app_secret)
        self._user_auth_token: Optional[str] = None

    def is_ready(self) -> bool:
        return self._ready

    def has_user_session(self) -> bool:
        return self._ready and bool(self._user_auth_token)

    def set_auth_token(self, token: str):
        self._user_auth_token = token.strip() or None

    async def initialize(self):
        if self._ready:
            return
        creds = await self._scrape_credentials()
        if creds:
            self.app_id, self.app_secret = creds
            self._ready = True

    async def _scrape_credentials(self) -> Optional[tuple[str, str]]:
        try:
            response = await self.client.get(CREDENTIAL_URL, headers={"User-Agent": "Mozilla/5.0"})
            response.raise_for_status()
            js_match = re.search(r'<script[^>]+src="([^"]+/js/main\.js)"', response.text)
            if not js_match:
                return None
            js_url = js_match.group(1)
            if js_url.startswith("/"):
                js_url = "https://open.qobuz.com" + js_url
            script = await self.client.get(js_url, headers={"User-Agent": "Mozilla/5.0"})
            script.raise_for_status()
            match = re.search(r'app_id:"(\d{9})",app_secret:"([a-f0-9]{32})"', script.text)
            return (match.group(1), match.group(2)) if match else None
        except httpx.HTTPError:
            return None

    def _sign(self, path: str, params: dict[str, str]) -> dict[str, str]:
        if not self._ready:
            raise RuntimeError("Qobuz application credentials are unavailable")
        timestamp = str(int(time.time()))
        path_norm = path.strip("/").replace("/", "")
        keys = sorted(key for key in params if key not in {"app_id", "request_ts", "request_sig"})
        payload = path_norm + "".join(key + params[key] for key in keys) + timestamp + self.app_secret
        signature = hashlib.md5(payload.encode(), usedforsecurity=False).hexdigest()
        return {
            **params,
            "app_id": self.app_id,
            "request_ts": timestamp,
            "request_sig": signature,
        }

    async def search(self, query: str, limit: int = 30) -> list[dict[str, Any]]:
        if not self._ready:
            return []
        params = self._sign("track/search", {"query": query, "limit": str(limit)})
        try:
            response = await self.client.get(SEARCH_URL, params=params, headers=self._headers())
            response.raise_for_status()
            data = response.json()
            raw_tracks = data.get("tracks", {})
            tracks = raw_tracks.get("items", []) if isinstance(raw_tracks, dict) else raw_tracks
            return [self._map_track(track) for track in tracks[:limit]]
        except (httpx.HTTPError, ValueError, TypeError):
            return []

    async def get_stream_url(self, track_id: str, quality: str = "24") -> Optional[str]:
        stream = await self.get_stream(track_id, quality)
        return str(stream["url"]) if stream else None

    async def get_stream(self, track_id: str, quality: str = "24") -> Optional[dict[str, Any]]:
        # A user session is required for full files. Public getFileUrl responses
        # are previews and must never be returned as playable tracks.
        if not self.has_user_session():
            return None
        format_ids = QUALITY_FORMATS.get(quality.strip().lower(), QUALITY_FORMATS["24"])
        for format_id in format_ids:
            result = await self._get_file_url(track_id, format_id)
            if result:
                return result
        return None

    async def _get_file_url(self, track_id: str, format_id: int) -> Optional[dict[str, Any]]:
        params = {
            "track_id": track_id,
            "format_id": str(format_id),
            "intent": "stream",
            "user_auth_token": self._user_auth_token or "",
        }
        try:
            response = await self.client.get(STREAM_URL, params=self._sign("track/getFileUrl", params), headers=self._headers())
            if response.status_code != 200:
                return None
            data = response.json()
            url = str(data.get("url") or data.get("file_url") or "").strip()
            if not self._is_full_https_stream(url):
                return None
            bit_depth = self._positive_int(data.get("bit_depth"))
            sample_rate_hz = self._sample_rate_hz(data.get("sampling_rate"))
            mime_type = str(data.get("mime_type") or ("audio/mpeg" if format_id == 5 else "audio/flac"))
            parsed_query = parse_qs(urlparse(url).query)
            expires_at = self._positive_int((parsed_query.get("etsp") or parsed_query.get("expires") or [None])[0])
            hi_res = bool((bit_depth and bit_depth > 16) or (sample_rate_hz and sample_rate_hz > 44_100))
            return {
                "provider": "qobuz",
                "url": url,
                "format": "mp3" if format_id == 5 else "flac",
                "mimeType": mime_type,
                "quality": "Hi-Res FLAC" if hi_res else ("MP3 320" if format_id == 5 else "16-bit / 44.1 kHz FLAC"),
                "bitDepth": bit_depth,
                "sampleRateHz": sample_rate_hz,
                "channelCount": 2,
                "expiresAt": expires_at,
                "isHiRes": hi_res,
                "isDolbyAtmos": False,
            }
        except (httpx.HTTPError, ValueError, TypeError):
            return None

    def _headers(self) -> dict[str, str]:
        return {
            "User-Agent": "Mozilla/5.0",
            "Accept": "application/json",
            "X-App-Id": self.app_id,
        }

    @staticmethod
    def _is_full_https_stream(url: str) -> bool:
        try:
            parsed = urlparse(url)
            if parsed.scheme != "https" or not parsed.hostname or parsed.username or parsed.password:
                return False
            query = {key.lower() for key in parse_qs(parsed.query)}
            return not bool(query & {"range", "preview", "sample", "range_start", "range_end"})
        except ValueError:
            return False

    @staticmethod
    def _positive_int(value: Any) -> Optional[int]:
        try:
            parsed = int(float(value))
            return parsed if parsed > 0 else None
        except (TypeError, ValueError):
            return None

    @classmethod
    def _sample_rate_hz(cls, value: Any) -> Optional[int]:
        rate = cls._positive_int(value)
        if not rate:
            return None
        return rate * 1000 if rate < 1000 else rate

    def _map_track(self, track: dict[str, Any]) -> dict[str, Any]:
        performer = track.get("performer", {}) or {}
        album = track.get("album", {}) or {}
        return {
            "id": str(track.get("id", "")),
            "title": track.get("title", ""),
            "artist": performer.get("name", ""),
            "album": album.get("title", ""),
            "duration": track.get("duration", 0),
            "isrc": track.get("isrc", ""),
            "artwork": self._get_artwork(track),
            "quality": "hi-res" if track.get("hires", False) else "lossless",
            "service": "qobuz",
        }

    @staticmethod
    def _get_artwork(track: dict[str, Any]) -> str:
        cover = (track.get("album", {}) or {}).get("cover", {}) or {}
        for size in ("large", "medium", "small", "original"):
            if cover.get(size):
                return str(cover[size])
        return ""
